File: common/src/main/java/dev/obscuria/fragmentum/util/easing/CubicCurve.java
package dev.obscuria.fragmentum.util.easing;

import net.minecraft.util.Mth;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;

@SuppressWarnings("unused")
public final class CubicCurve
{
    private final List<Point> controlPoints = new ArrayList<>();
    private final List<Point> points = new ArrayList<>();
    private final int resolution;
    private boolean isDirty = true;

    public CubicCurve(int resolution)
    {
        this.resolution = Mth.clamp(resolution, 0, 100);
    }

    public CubicCurve addPoint(double x, double y)
    {
        return addPoint((float) x, (float) y);
    }

    public CubicCurve addPoint(float x, float y)
    {
        controlPoints.add(new Point(x, y));
        controlPoints.sort(Comparator.comparingDouble(it -> it.x));
        isDirty = true;
        return this;
    }

    public Point samplePoint(float x)
    {
        if (isDirty) bake();
        if (points.isEmpty()) return Point.ZERO;
        final var index = Mth.clamp(x * (points.size() - 1), 0, points.size() - 1);
        return points.get((int) index);
    }

    public float sample(float x)
    {
        return samplePoint(x).y;
    }

    public EasingFunction toEasing()
    {
        return this::sample;
    }

    private void bake()
    {
        isDirty = false;
        points.clear();

        final var res = 1f / (float) resolution;
        final var segments = controlPoints.size() - 1;

        for (int i = 0; i < segments; i++)
        {
            final var point = controlPoints.get(i);
            final var nextPoint = controlPoints.get(i + 1);

            final var xDiff = nextPoint.x - point.x;
            final var yDiff = nextPoint.y - point.y;

            var step = 0f;
            while (step <= 1f)
            {
                float x = point.x + step * xDiff;
                float y = point.y + step * step * (3 * yDiff - 2 * step * yDiff);
                points.add(new Point(x, y));
                step += res;
            }
        }
    }

    public record Point(float x, float y)
    {
        public static final Point ZERO = new Point(0, 0);
    }
}

File: common/src/main/kotlin/dev/obscuria/fragmentum/tools/easing/CubicCurve.kt
@file:Suppress("unused", "UNUSED_PARAMETER", "RedundantNullableReturnType", "MemberVisibilityCanBePrivate")

package dev.obscuria.fragmentum.tools.easing

import net.minecraft.util.Mth
import kotlin.math.max
import kotlin.math.min

class CubicCurve(resolution: Int) {

    private val rawPoints: MutableList<Point> = mutableListOf()
    private val cachedPoints: MutableList<Point> = mutableListOf()
    private val resolution = min(max(resolution.toDouble(), 0.0), 100.0).toInt()
    private var dirty = true

    fun addPoint(x: Double, y: Double): CubicCurve {
        return this.addPoint(x.toFloat(), y.toFloat())
    }

    fun addPoint(x: Float, y: Float): CubicCurve {
        rawPoints.add(Point(Mth.clamp(x, 0f, 1f), Mth.clamp(x, 0f, 1f)))
        rawPoints.sortWith(Comparator.comparingDouble { it.x.toDouble() })
        this.dirty = true
        return this
    }

    fun samplePoint(x: Float): Point {
        if (this.dirty) this.bake()
        if (cachedPoints.isEmpty()) return Point.ZERO
        val index = Mth.clamp(x * (cachedPoints.size - 1), 0f, (cachedPoints.size - 1).toFloat()).toInt()
        return cachedPoints[index]
    }

    fun sample(x: Float): Float {
        return samplePoint(x).y
    }

    fun asEasing(): EasingFunction {
        return EasingFunction { progress -> sample(progress) }
    }

    private fun bake() {
        this.dirty = false
        cachedPoints.clear()

        val res = 1f / resolution.toFloat()
        val segments = rawPoints.size - 1

        for (i in 0..<segments) {
            val point: Point = rawPoints[i]
            val nextPoint: Point = rawPoints[i + 1]
            val xDiff: Float = nextPoint.x - point.x
            val yDiff: Float = nextPoint.y - point.y

            var step = 0f
            while (step <= 1) {
                val x: Float = point.x + step * xDiff
                val y: Float = point.y + step * step * (3 * yDiff - 2 * step * yDiff)
                cachedPoints.add(Point(x, y))
                step += res
            }
        }
    }

    data class Point(val x: Float, val y: Float) {

        companion object {

            val ZERO: Point = Point(0f, 0f)
        }
    }
}